algebra/vector_space.lean

```diff
@@ -0,0 +1,305 @@
+import algebra.ring
+import algebra.module
+
+universes u' v'
+
+class vector_space (α : inout Type u')  (β : Type v') [field α] extends module α β
+
+instance field.to_vector_space {K : Type} [field K] :
+vector_space K K := ⟨ring.to_module⟩
+
+lemma eq_zero_of_add_self_eq {α : Type} [add_group α]
+{a : α} (H : a + a = a) : a = 0 :=
+add_left_cancel (by {rw add_zero, exact H})
+
+class subspace (K : Type) (V : Type) [field K] [vector_space K V] (p : V → Prop) :=
+    (p_zero : p 0)
+    (p_add : ∀ u v, p u → p v → p (u + v))
+    (p_neg : ∀ v, p v → p (-v))
+    (p_smul : ∀ c v, p v → p (c • v))
+
+namespace subspace
+
+variables (K : Type) {V : Type} [field K] [vector_space K V]
+{p : V → Prop} [subspace K V p]
+
+variables {c d : K} (u v : {x // p x})
+
+include K
+
+def add (u v : {x // p x}) : {x // p x} :=
+{
+    val := u.val + v.val,
+    property := p_add K u.val v.val u.property v.property
+}
+
+def zero : {x // p x} :=
+{
+    val := 0,
+    property := p_zero K p
+}
+
+def neg (v : {x // p x}) : {x // p x} :=
+{
+    val := -v.val,
+    property := p_neg K v.val v.property
+}
+
+def smul (c : K) (v : {x // p x}) : {x // p x} :=
+{
+    val := c • v.val,
+    property := p_smul c v.val v.property
+}
+
+instance : has_add {x // p x} := ⟨add K⟩
+instance : has_zero {x // p x} := ⟨zero K⟩
+instance : has_neg {x // p x} := ⟨neg K⟩
+instance : has_scalar K {x // p x} := ⟨smul K⟩
+
```

Comment:
same here as with `units` (a.k.a. `invertibles`): avoid `zero`, `neg`, `smul`... and use `has_zero`, etc directtly.